scene/3d/sprite_3d.cpp

```diff
@@ -504,25 +550,61 @@ void Sprite3D::_draw() {
 
 	AABB aabb;
 
-	for (int i = 0; i < 4; i++) {
-		VS::get_singleton()->immediate_normal(immediate, normal);
-		VS::get_singleton()->immediate_tangent(immediate, tangent);
-		VS::get_singleton()->immediate_color(immediate, color);
-		VS::get_singleton()->immediate_uv(immediate, uvs[i]);
+	// Buffer is using default compression, so everything except position is compressed
+	PoolVector<uint8_t>::Write write_buffer = mesh_buffer.write();
 
+	int8_t v_normal[4] = {
+		(int8_t)CLAMP(normal.x * 127, -128, 127),
+		(int8_t)CLAMP(normal.y * 127, -128, 127),
+		(int8_t)CLAMP(normal.z * 127, -128, 127),
+		0,
+	};
+
+	int8_t v_tangent[4] = {
+		(int8_t)CLAMP(tangent.normal.x * 127, -128, 127),
+		(int8_t)CLAMP(tangent.normal.y * 127, -128, 127),
+		(int8_t)CLAMP(tangent.normal.z * 127, -128, 127),
+		(int8_t)CLAMP(tangent.d * 127, -128, 127)
+	};
+
+	uint8_t v_color[4] = {
+		(uint8_t)CLAMP(int(color.r * 255.0), 0, 255),
+		(uint8_t)CLAMP(int(color.g * 255.0), 0, 255),
+		(uint8_t)CLAMP(int(color.b * 255.0), 0, 255),
+		(uint8_t)CLAMP(int(color.a * 255.0), 0, 255)
+	};
+
+	for (int i = 0; i < 4; i++) {
 		Vector3 vtx;
 		vtx[x_axis] = vertices[i][0];
 		vtx[y_axis] = vertices[i][1];
-		VS::get_singleton()->immediate_vertex(immediate, vtx);
 		if (i == 0) {
 			aabb.position = vtx;
 			aabb.size = Vector3();
 		} else {
 			aabb.expand_to(vtx);
 		}
+
+		uint16_t v_uv[2] = { Math::make_half_float(uvs[i].x), Math::make_half_float(uvs[i].y) };
+		copymem(&write_buffer[i * mesh_stride + mesh_surface_offsets[VS::ARRAY_VERTEX]], &vertices[i], sizeof(float) * 2);
+		copymem(&write_buffer[i * mesh_stride + mesh_surface_offsets[VS::ARRAY_NORMAL]], v_normal, 4);
+		copymem(&write_buffer[i * mesh_stride + mesh_surface_offsets[VS::ARRAY_TANGENT]], v_tangent, 4);
+		copymem(&write_buffer[i * mesh_stride + mesh_surface_offsets[VS::ARRAY_COLOR]], v_color, 4);
+		copymem(&write_buffer[i * mesh_stride + mesh_surface_offsets[VS::ARRAY_TEX_UV]], v_uv, 2 * 2);
```

Comment:
Seeing the half float use, can Sprite3D deal with situation when half float is not supported? (maybe it's hard coded to half float somewhere else, but thought this might be worth mentioning).

Reply:
This is half float because meshes use half_float for uvs by default. I think this is handled well enough inside rasterizer_storage.

In GLES3 it should always be fine. In GLES2 the ``_unpack_half_float()`` function should be enough.

https://github.com/godotengine/godot/blob/1ebfe84458373fbddf4e2b3b58d58e6a858fc662/drivers/gles2/rasterizer_storage_gles2.cpp#L2055